src/HydraAuctionOnchain/MintingPolicies/Auction.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module HydraAuctionOnchain.MintingPolicies.Auction
+  ( auctionMintingPolicy
+  ) where
+
+import HydraAuctionOnchain.Errors.MintingPolicies.Auction (PAuctionMpError (..))
+import HydraAuctionOnchain.Helpers (pdecodeInlineDatum, pfindUniqueOutputWithScriptHash)
+import HydraAuctionOnchain.Lib.ScriptContext (pownCurrencySymbol)
+import HydraAuctionOnchain.Types.AuctionInfo (PAuctionInfo)
+import HydraAuctionOnchain.Types.AuctionTerms (pvalidateAuctionTerms)
+import HydraAuctionOnchain.Types.Error (errCode, passert, passertMaybe)
+import HydraAuctionOnchain.Types.Tokens
+  ( pauctionTokenBundleBurned
+  , pauctionTokenBundleMinted
+  , ptxOutContainsAuctionMetadataToken
+  )
+import Plutarch.Api.V1.Value (pnormalize)
+import Plutarch.Api.V2 (PCurrencySymbol, PScriptContext, PScriptHash, PTxOutRef)
+import Plutarch.Extra.Maybe (pisJust)
+import Plutarch.Extra.ScriptContext (pfindTxInByTxOutRef)
+import Plutarch.Monadic qualified as P
+
+data PAuctionMpRedeemer (s :: S)
+  = MintAuctionRedeemer (Term s (PDataRecord '[]))
+  | BurnAuctionRedeemer (Term s (PDataRecord '[]))
+  deriving stock (Generic)
+  deriving anyclass (PlutusType, PIsData, PShow, PEq)
+
+instance DerivePlutusType PAuctionMpRedeemer where
+  type DPTStrat _ = PlutusTypeData
+
+auctionMintingPolicy
+  :: Term
+      s
+      ( PScriptHash
+          :--> PTxOutRef
+          :--> PAuctionMpRedeemer
+          :--> PScriptContext
+          :--> PUnit
+      )
+auctionMintingPolicy = phoistAcyclic $
+  plam $ \auctionMetadataSh nonceOref redeemer ctx -> P.do
+    -- Script purpose should contain own currency symbol. Always
+    -- true for minting policies.
+    auctionCs <-
```

Comment:
Out of curiosity, I wonder how it is costlier than `PMinting m <- pmatch $ ctxFs.purpose`